### ISSUE TITLE
[Feature Request] Hexes in Range 

### ISSUE THREAD

**Issue by grandsong** [flauwekeul/honeycomb#17]
Simply put, I hope you kindly implement [ redblobgames.com # range-coordinate ](https://www.redblobgames.com/grids/hexagons/#range-coordinate).
It is very useful and of great demand.
I can implement this myself, but I don't like hacking such a good lib like yours... And I guess other people also need this.

**Comment by flauwekeul** [comment]
Yes! I've always wanted to implement a feature request from one of honeycomb's users 😄

I'll get started on this 👍

**Comment by flauwekeul** [comment]
I've added an instance method `hexesInRange()` to `Grid`, with the API below. I haven't released it yet. Please let me know if this does what you need and I'll create a new minor release.

### Parameters
**`centerHex`** (`hex`) A hex to get surrounding hexes from.
**`range`** (`number = 0`) The range (in hexes) surrounding the center hex.

### Returns
`Array<hex>`: An array with all hexes surrounding the passed center hex. Only hexes that are present in the grid are returned.

### Throws
Error: When no valid hex is passed.

### Example
```js
const Hex = Honeycomb.extendHex({ orientation: 'pointy' })
const Grid = Honeycomb.defineGrid(Hex)
const grid = Grid.rectangle({ width: 5, height: 5 })

grid.hexesInRange(Hex(2, 2), 2)      // [
                                     //    { x: 0, y: 2 },
                                     //    { x: 0, y: 3 },
                                     //    { x: 1, y: 4 },
                                     //    ...
                                     //    { x: 3, y: 0 },
                                     //    { x: 3, y: 1 },
                                     //    { x: 4, y: 2 }
                                     // ]

// only returns hexes that exist in the grid:
grid.hexesInRange(Hex(-1, -1), 2)    // [
                                     //    { x: 0, y: 0 },
                                     //    { x: 0, y: 1 },
                                     //    { x: 1, y: 0 },
                                     // ]
```

**Comment by flauwekeul** [comment]
I like the idea of an option to include/exclude the center hex, I'll add it.

Well, not be a smart-ass, but why **not** accept a center hex that isn't present in the grid? The alternative is to throw an error, or maybe fail silently and not return anything.

**Comment by grandsong** [comment]
It's just about consistency. The returned results are all present in the grid, so should the center one.
And what does it mean for result hexes to "be in the range of" a center hex if that "center" is actually not in the grid?

**Comment by grandsong** [comment]
Some games have irregular hex map where "holes" and "gaps" exists between hexes.
Thoes "holes" and "gaps" have their potions / point, but nothing can sit on / move to them.
If you throw a grenade, you cannot throw it to a hole. It would be meaningless to allow players to do so.
As a result, even when a bunch of enemy units are around a hole, you cannot blow them by targeting that hole.
--
In some games, "holes" can be moved to but the poor unti will fall... Well, such holes are in fact present in the grid, just of a special role and with special functions. This is another story.

**Comment by flauwekeul** [comment]
I agree it might be a little unexpected to allow a non-existent center hex. But I don't see the harm in allowing it either.

I can imagine a scenario where a player throws a grenade outside a grid, but that grenade could still do damage in the grid. The grid could represent where the player may move, but the world doesn't stop existing outside this grid. There might be other hexes, unable for the player to move, but able to "catch" a thrown grenade...

Currently the "rule" is: grid methods always return hexes that are present in the grid and hex methods always just return hexes (regardless of any grid). In a future version I'd like to have a more explicit way of filtering returned hexes by "grid-presence". Not sure how yet...

**Comment by grandsong** [comment]
Again, your assumption of the uses of hexes and grids falls in the category like playlist and songs, which I highly doubt is reasonable.

Your imagined case is exactly what I meant "such holes are in fact present in the grid": they are visually holes, but they are present.

When we talk about "present", I think the concerns are actually about relations.

Hex worlds are complex. Let me pick examples from much simpler ones, such as arrays.

If an array `entities` is `['hero', undefined, undefined, 'slime']`, the second cell is not present.
So, the hero (at index 0) cannot interact with this position (index 1) because nothing is there.

But this doesn't look right. (What about movement? Or, grenade?)
No, it isn't right.

When we use matrix-like structures (array is one-dimensional, while hex is... whatever), they are best suitble for **continuum**.

Yeah, like maps.

---

There are two kinds of maps:

1) Completely continuous. No "holes" at all.
2) Partly continuous. (If you know games like [Bastion](https://store.steampowered.com/app/107100/Bastion), you will have a clear idea about what these maps are like)

But, when rendering a game scene, there is always a grid of completely continuous tiles to "contain" the entities when needed.
Terrain/ground cells in maps are just of one layer.

---

Back to the array example. There should be more arrays to make a scene work.

1) tiles: `[0, 1, 2]`
2) terrains: `['plain', 'plain', 'wall', 'plain']`
3) fixures: `['grass', 'tree', null, null]`
4) moveable objects and characters: well, it depends...

If characters always move like chess pieces, and they are many, it is reasonable to use array like preview example.
But usually, it is not the case.
Even chess pieces are recorded independently...

---

Let's move on to a non-game example: Excel.

When we create a spreadsheet, we see a very, very huge grid. The count of rows is the limit of long integer.
But this grid is just to "contain" the incoming data. Without data, when saved, the spreadsheet will be blank.

If you cut and paste (or, move) a range of cells, you actually move the data, not the cells in the grid: those cells where the data were are now empty.
And you can place several table of data, or any irregular group of data, in one spreadsheet.
The grid itself is always intact. Cells are just "seats" for data to sit in.
In Excel files (zipped xml), all data are recorded independently, not in a matrix.

---

In game "Prison Architect" (very very successful, and I'm a big fan of it), for a prison, all things are saved in one huge (the size of a prison can be eg 400 * 500 square tiles) text (yeah, not crypt nor compressed) file, in the fashion similar to Excel xml format.
The sqaure grid is not stored. It is created in the runtime.

Players can pay and expand the size of a map.
If the map expands up, the (0, 0) tile is actually "moved", then the positions of all entities are automatically changed with offset to fit the new coordinate system.

---

Let me sum up.
There are several levels of things overlap:

#### coordinated space

which is **pure mathmatical** and can be **virtually infinite**.

#### scene scope

which has a finite size.

The tiles in (as content of)the map are always continual. But they can be "empty".
In fact, when initiating a map, all tiles should be fed with data, otherwise they are empty.

Positions out of the scope is meaningless and should be taken as errors.
Exceptions can be allowed as long as the designer/developer knows why there need to be exceptions.

#### entities

They are not tiles.
No matter whether they look like ones (eg walls look like square tiles).
No matter whether they always take (occupy) tiles and move at steps of tiles.

---

The most common uses of grid is scene scope.

Some map editors and Excel use coordinated space because they don't know how big is big enough.
Other map editors require user to set sizes of the map.

In games, map is usually fixed.

Entities are placed over tiles. Entities don't compose grids.
My earlier words about 5 + x layers were incorrect.

**Comment by flauwekeul** [comment]
I've changed the method to throw an error when a non-existing hex is passed and released v1.4.

I can follow most things you're saying. The `Grid` factory is a bit like Excel: an infinite grid of defined hexes. A grid instance is a concrete finite grid of the same defined hexes.

I'll close this issue because the requested feature is added.

**Comment by grandsong** [comment]
Good!
